sno/gpkg.py

```diff
@@ -317,25 +294,95 @@ def hex_wkb_to_gpkg_geom(hex_wkb, **kwargs):
 WKB_POINT_EMPTY_LE = b"\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xF8\x7F\x00\x00\x00\x00\x00\x00\xF8\x7F"
 
 
+def _gpkg_envelope_size(flags):
+    envelope_typ = (flags & 0b00001110) >> 1
+    if envelope_typ == 1:
+        return 32
+    elif envelope_typ in (2, 3):
+        return 48
+    elif envelope_typ == 4:
+        return 64
+    elif envelope_typ == 0:
+        return 0
+    else:
+        raise ValueError("Invalid envelope contents indicator")
+
+
+def _wkb_endianness_and_geometry_type(buf, wkb_offset=0):
+    """
+    Given a buffer containing some WKB at the given offset,
+    returns a two-tuple:
+        * is_little_endian (bool)
+        * geometry type as an integer.
+    """
+    (is_le,) = struct.unpack_from('b', buf, offset=wkb_offset)
+    (typ,) = struct.unpack_from(f'{"<" if is_le else ">"}I', buf, offset=wkb_offset + 1)
+    return is_le, typ
+
+
+def normalise_gpkg_geom(gpkg_geom):
+    """
+    Checks to see if the given gpkg geometry:
+        * is little-endian
+        * has little-endian WKB
+        * has an envelope
+
+    If so, returns the geometry unmodified.
+    Otherwise, returns a little-endian geometry with an envelope attached.
+    """
+    if gpkg_geom is None:
+        return None
+    flags = _validate_gpkg_geom(gpkg_geom)
+
+    # http://www.geopackage.org/spec/#flags_layout
+    is_le = bool(flags & 0x0001) != 0
+    want_envelope = None
+    if is_le:
+        envelope_size = _gpkg_envelope_size(flags)
+        has_envelope = bool(envelope_size)
+
+        # is this a point? if so, we don't *want* an envelope
+        wkb_offset = 8 + envelope_size
+        wkb_is_le, geom_type = _wkb_endianness_and_geometry_type(
+            gpkg_geom, wkb_offset=wkb_offset
+        )
+        want_envelope = geom_type != ogr.wkbPoint
+
+        if wkb_is_le and has_envelope == want_envelope:
+            # everything is fine, no need to roundtrip via OGR
```

Comment:
what about empty geometry collections (or geometries)? Presumably they don't need envelopes either?

Reply:
correct, I guess I can exclude them. I don't want the logic for this to get _complicated_ but "empty or point" isn't too bad

Reply:
I think since this is used for hashing we need to get it definitively one way, and it's not going to be trivial to change

Reply:
From https://github.com/koordinates/sno/pull/224#pullrequestreview-472303962 (but that's me writing down what I remember/made sense, not canonical):

1.  So either we go with "Envelopes must be populated for all non-empty geometries except for `POINT[ZM]` types." ie:

- `POINT(0 0)` → no
- `MULTIPOINT ((0 0))` → yes*
- `MULTIPOINT ((0 0), (1 1))` → yes
- `GEOMETRYCOLLECTION(POINT(0 0))` → yes*
- `GEOMETRYCOLLECTION(MULTIPOINT(0 0))` → yes*
- `GEOMETRYCOLLECTION(POINT(0 0), POINT(1 1))` → yes

where yes* means (bbox top==bottom, left==right)

Or technically "better" but maybe slightly more complex to implement — "Envelopes must be populated for all geometries containing less than two vertices. In practise this means non-empty geometries and single POINT geometries". ie:

- `POINT(0 0)` → no
- `MULTIPOINT ((0 0))` → no
- `MULTIPOINT ((0 0), (1 1))` → yes
- `GEOMETRYCOLLECTION(POINT(0 0))` → no
- `GEOMETRYCOLLECTION(MULTIPOINT(0 0))` → no
- `GEOMETRYCOLLECTION(POINT(0 0), POINT(1 1))` → yes



2. Envelopes must have the same number of dimensions as the geometry (ie: a `POLYGONZM` geometry shall have a 4 dimensional envelope). For a `GEOMETRYCOLLECTION` the envelope shall have a dimensionality equal to the maximum dimensionality of the collected geometries.

- `GEOMETRYCOLLECTION(POINTZM(0 0), LINESTRING(1 1, 2 2))` → 2D
- `GEOMETRYCOLLECTION(POINTZM(0 1 2 3), LINESTRING(1 1, 2 2))` → 4D

Reply:
> Or technically "better" but maybe more complex to implement

definitely too complex. The optimisation for points was due to not wanting to double the storage size for a whole entire points layer (which are really _common_) and because it's fast to parse a point and simple to reason about. Niche cases like geometry collections with one point in them are going to break this simplicity.

What we don't want is to have to do serious work to figure out whether a geometry should or should not have an envelope added.

Reply:
> GEOMETRYCOLLECTION(POINTZM(0 1 2 3), LINESTRING(1 1, 2 2)) → 4D


🤯 I had no idea geoemtrycollection could contain geometries with differing dimensionality.
At present this PR always just adds 2D envelopes. I'll try and make it add 3d/4d ones.

Reply:
> definitely too complex

ok, so all the cases in (1), and that definition into the dataset format documentation.

> I had no idea geoemtrycollection could contain geometries with differing dimensionality.

they can contain anything 😄 

> At present this PR always just adds 2D envelopes. I'll try and make it add 3d/4d ones.

ya, lets not hobble ZM geometries for no real reason.

Reply:
> no real reason

It's going to mean a sizable slowdown here: https://github.com/koordinates/sno/blob/geom-bboxes/sno/geometry.py#L124-L142

because we need to parse the WKB of every geometry to determine what size envelope to give it. Including iterating over all subcomponents of collections.

I don't _really_ want to implement wkb parsing myself, so essentially it negates all the shortcuts there, and the whole normalise method becomes a roundtrip via `ogr_to_gpkg_geom(gpkg_geom_to_ogr(g))`.

Reply:
I'll merge this PR because it's growing too many limbs, but if we need to tackle 3d/4d envelopes I can do that in a followup.